File: ga/glm4_result_2024_6/result/1968D_glm/13_1968D.c
use std::io;
use std::cmp::{min, max};

const N: usize = 200000;

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read line");
    let t: usize = input.trim().parse().expect("Invalid input");

    for _ in 0..t {
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");
        let mut values = input.trim().split_whitespace().map(|s| s.parse::<usize>().unwrap());
        let (n, k, a, b) = (values.next().unwrap(), values.next().unwrap(), values.next().unwrap(), values.next().unwrap());
        let m = min(n, k);
        let mut arr = vec![0; n + 1];
        let mut aa = vec![0; n + 1];

        for x in 1..=n {
            arr[x] = values.next().unwrap();
        }
        for x in 1..=n {
            aa[x] = values.next().unwrap();
        }

        let mut p1 = 0;
        let mut p1_ = 0;
        let mut p2 = 0;
        let mut p2_ = 0;
        let mut a = a;
        let mut b = b;

        for x in 1..=m {
            p1 = max(p1, p1_ + aa[a] * (k - x + 1));
            p1_ += aa[a];
            a = arr[a];
            p2 = max(p2, p2_ + aa[b] * (k - x + 1));
            p2_ += aa[b];
            b = arr[b];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}

File: ga/4omini_result_2024_7/result/1968D_cpp_mini/13_1968D.rs
use std::io;

fn calculate(p: &Vec<usize>, a: &Vec<i64>, mut pos: usize, mut k: i64) -> i64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];

    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = mx.max(cur + k * a[pos]);
        cur += a[pos];
        pos = p[pos];
        k -= 1;
    }
    
    mx
}

fn main() {
    let stdin = io::stdin();
    let mut input = String::new();
    stdin.read_line(&mut input).unwrap();
    let t: i32 = input.trim().parse().unwrap();
    
    for _ in 0..t {
        input.clear();
        stdin.read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace();
        let n: usize = iter.next().unwrap().parse().unwrap();
        let k: i64 = iter.next().unwrap().parse().unwrap();
        let ps1: usize = iter.next().unwrap().parse().unwrap() - 1;
        let ps2: usize = iter.next().unwrap().parse().unwrap() - 1;

        let mut p = Vec::with_capacity(n);
        let mut a = Vec::with_capacity(n);

        input.clear();
        stdin.read_line(&mut input).unwrap();
        for num in input.split_whitespace() {
            p.push(num.parse::<usize>().unwrap() - 1);
        }

        input.clear();
        stdin.read_line(&mut input).unwrap();
        for num in input.split_whitespace() {
            a.push(num.parse::<i64>().unwrap());
        }

        let ans1 = calculate(&p, &a, ps1, k);
        let ans2 = calculate(&p, &a, ps2, k);

        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}